internal/controller/engine/engine.go

```diff
@@ -0,0 +1,434 @@
+		}
+
+		// We've already created this watch and the informer backing it is still
+		// running. We don't need to create a new watch.
+		if watchExists[gvk] && activeInformer[gvk] {
+			e.log.Debug("Watch exists for GVK, not starting a new one", "controller", name, "watched-gvk", gvk)
+			continue
+		}
+
+		start[gvk] = w
+	}
+
+	// Don't take any write locks if there's no watches to start.
+	if len(start) == 0 {
+		return nil
+	}
+
+	// TODO(negz): If this blocks too much, we could alleviate it a bit by
+	// reading watches to start from a buffered channel.
+
+	// Take the write lock. This will block any other callers that want to
+	// update the watches for the same controller.
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	// Start new sources.
+	for gvk, w := range start {
+		// The controller's Watch method just calls the StoppableSource's Start
+		// method, passing in its private work queue as an argument. This
+		// will start an informer for the watched kind if there isn't one
+		// running already.
+		//
+		// The watch will stop sending events when either the source is stopped,
+		// or its backing informer is stopped. The controller's work queue will
+		// stop processing events when the controller is stopped.
+		src := NewStoppableSource(e.infs, w.kind, w.handler, w.predicates...)
+		if err := c.ctrl.Watch(src); err != nil {
+			return errors.Wrapf(err, "cannot start watch for %q", gvk)
+		}
+
+		// Record that we're now running this source.
+		c.sources[gvk] = src
+
+		e.log.Debug("Started watching GVK", "controller", name, "watched-gvk", gvk)
+	}
+
+	return nil
+}
+
+// StopWatches stops all watches for the supplied controller, except watches for
+// the GVKs supplied by the keep argument. StopWatches blocks other operations
+// on the same controller if and when it stops a watch. It returns the number of
+// watches that it successfully stopped.
+func (e *ControllerEngine) StopWatches(ctx context.Context, name string, keep ...schema.GroupVersionKind) (int, error) {
+	e.mx.RLock()
+	c, ok := e.controllers[name]
+	e.mx.RUnlock()
+
+	if !ok {
+		return 0, errors.Errorf("controller %q is not running", name)
+	}
+
+	c.mx.RLock()
+	stop := sets.KeySet(c.sources).Difference(sets.New(keep...))
+	c.mx.RUnlock()
+
+	// Don't take the write lock if we actually want to keep all watches.
+	if len(stop) == 0 {
+		return 0, nil
+	}
+
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	stopped := 0
+	for gvk := range stop {
+		if err := c.sources[gvk].Stop(ctx); err != nil {
+			return stopped, errors.Wrapf(err, "cannot stop watch for %q", gvk)
+		}
+		delete(c.sources, gvk)
+		e.log.Debug("Stopped watching GVK", "controller", name, "watched-gvk", gvk)
+		stopped++
+	}
+
+	return stopped, nil
+}
+
+// RemoveUnwatchedInformers removes all informers that don't power any
+// controller's watches. This may include informers that weren't started by a
+// watch, e.g. informers that were started by a Get or List call to a cache
```

Comment:
are you saying nobody else should use this cache because then a regular `Get` or `List` can make the cache start and stop informers continuously?

Reply:
Yes. There's a comment to that effect inside `type ControllerEngine struct` - that the engine must have exclusive use of its tracking cache. I should add it to the `New` docstring too.

When I wrote this my thought was:

* If realtime compositions are enabled, we'll always have a watch for any type we read (e.g. via Get or List) so we can use presence of sources to determine whether an informer is used or unused.
* If realtime compositions aren't enabled, we'll never start informers for composed resources (the default client behaviour is to not start informers for unstructured resources on Get or List).

However, thinking through the controllers managed by the engine I missed some cases:

* I expect the claim controller to read claims, XRs, and secrets.
* I expect the XR controller to read XRs, compositions, composition revisions, environment configs, composed resources, and secrets.

We never cache secrets, so they're fine. I forgot about compositions and environment configs though. Nothing takes a watch on them and they're not unstructured, so I think their informers would "flap". We'd continually start one on Get/List and then remove it within 60 seconds.

Perhaps a TTL based approach would be a better way to GC informers.

Reply:
I believe this is the only blocker left on this PR - apart from adding tests for the new engine.

Three ways to address this come to mind:

* Use the manager's client to get Compositions or EnvironmentConfigs.
* Garbage collect informers based on a TTL that resets on each use.
* Garbage collect informers only when the CRD that defines the type they're watching is deleted.

Using the manager's client is trivial because we already support passing a specific client for these tasks when constructing the XR reconciler. My concern with this approach is that it seems hard to reason about - potentially error prone for future changes. When constructing a reconciler for a dynamic controller (i.e. a claim or XR controller) you'd need to use the engine's client for anything the controller might watch. You would instead use the manager's client for anything the controller doesn't watch.

My concern with the TTL based approach is that I believe it would require taking a write lock any time an informer was used, in order to update the TTL. So any cache read (i.e. call to `Get` or `List`) would block all other cache reads from all other controllers run by the engine.

My concern with the CRD based approach is that you could end up with unused informers wasting compute resources if you composed a bunch of resource types, then stopped composing them but didn't uninstall their provider. I could be convinced this concern isn't a serious one in practice, though.

Reply:
I pushed a commit with the CRD based approach. It feels simple and clean.

For now, I'm leaning toward not doing anything to cleanup CR informers that are unused but not deleted because their CRDs still exist. I can think of a few ways to address it but I'd rather wait to see whether it's a problem in practice.